stdlib/runtime/malloc.gr

```diff
@@ -0,0 +1,275 @@
+/* compilation-mode: malloc */
+
+import WasmI32, { 
+  add as (+),
+  sub as (-),
+  divU as (/),
+  gtU as (>),
+  geU as (>=),
+  ltU as (<),
+  shl as (<<),
+  shrU as (>>),
+  eq as (==),
+  ne as (!=)
+} from "runtime/unsafe/wasmi32"
+
+primitive (!) : Bool -> Bool = "@not"
+primitive (&&) : (Bool, Bool) -> Bool = "@and"
+primitive (||) : (Bool, Bool) -> Bool = "@or"
+
+import foreign wasm _malloc: (WasmI32, WasmI32) -> WasmI32 as extMalloc from "memoryManager"
+import foreign wasm _free: (WasmI32) -> WasmI32 as extFree from "memoryManager"
+import foreign wasm _growHeap: (WasmI32) -> WasmI32 as extGrowHeap from "memoryManager"
+import foreign wasm _initialHeapSize: WasmI32 as initialHeapSize from "memoryManager"
+
+/* UNDERSTANDING THE STRUCTURE OF THE FREE LIST
+ * The original K&R definition for the free list entry type was the following:
+ *
+ *     union header {
+ *         struct {
+ *             union header *ptr;
+ *             unsigned size;
+ *         } s;
+ *         long x; // <- forces 8-byte alignment
+ *     };
+ *
+ * In memory, this is really just two ints (assuming we're working in 32-bit mode).
+ * As such, we manually lay out the entries on the heap as follows (note that we
+ * use helpers to facilitate accessing and setting these values):
+ *
+ *   Grain             C Equivalent
+ *   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *   let ptr      ===  union header *ptr
+ *   getNext(ptr) ===  ptr->s.ptr
+ *   getSize(ptr) ===  ptr->s.size
+ */
+
+/**
+ * Pointer to the start of the free list. This is always a multiple of
+ * 8, with the exception of its initial value (used as a sentinel).
+ */
+let mut freePtr = 1n
+
+/**
+ * Size (in bytes) of entries in the free list.
+ */
+let mallocHeaderSize = 8n
+
+/**
+ * log_2(mallocHeaderSize) (multiplication by the header
+ * size is equivalent to left-shifting by this amount)
+ */
+let logMallocHeaderSize = 3n
+
+/**
+ * The current size (in bytes) of the heap.
+ */
+let mut heapSize = 0n
+
+let _RESERVED_RUNTIME_SPACE = 4096n;
+
+let getNext = (ptr: WasmI32) => {
+  WasmI32.load(ptr, 0n)
+}
+
+let setNext = (ptr: WasmI32, val: WasmI32) => {
+  WasmI32.store(ptr, val, 0n)
+}
+
+let getSize = (ptr: WasmI32) => {
+  WasmI32.load(ptr, 4n)
+}
+
+let setSize = (ptr: WasmI32, val: WasmI32) => {
+  WasmI32.store(ptr, val, 4n)
+}
```

Comment:
It could maybe be more clear that way, though I felt it was unnecessary since they're contained within these functions. But happy to change it if you think it'd be better!

Reply:
Looks like this was done in your updates.